File: src/main/java/io/github/binout/soccer/infrastructure/mail/SendGridMailService.java
package io.github.binout.soccer.infrastructure.mail;

import feign.*;
import io.github.binout.soccer.domain.notification.MailService;
import io.github.binout.soccer.infrastructure.log.LoggerService;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Arrays;

@Component
public class SendGridMailService implements MailService {

    private static final String SENDGRID_API = "https://api.sendgrid.com/v3";

    private final LoggerService loggerService;

    @Autowired
    public SendGridMailService(LoggerService loggerService) {
        this.loggerService = loggerService;
    }

    @Override
    public void sendMail(Mail email) {
        String apiKey = System.getenv("SENDGRID_API_KEY");
        if (apiKey != null) {
            if (email.hasRecipients()) {
                String jsonMail = toSendGridMail(email);
                try {
                    Feign.builder()
                            .requestInterceptor(r -> r.header("Authorization", "Bearer " + apiKey))
                            .target(SendGrid.class, SENDGRID_API)
                            .sendMail(jsonMail);
                } catch (Throwable e) {
                    loggerService.log(this.getClass(), "Cannot send mail to sendgrid api : " + jsonMail);
                    loggerService.log(this.getClass(), Arrays.toString(e.getStackTrace()));
                }
            }
        } else {
            loggerService.log(this.getClass(), "No SendGrid Configuration");
        }
    }

    static String toSendGridMail(Mail email) {
        try {
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("from", toEmail(email.from()));
            jsonObject.put("personalizations", toPersonalizations(email));
            jsonObject.put("content", toContent(email));
            return jsonObject.toString();
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    private static JSONArray toContent(Mail email) {
        try {
            JSONArray jsonArray = new JSONArray();
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("type", "text/html");
            jsonObject.put("value", email.content());
            jsonArray.put(jsonObject);
            return jsonArray;
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    private static JSONArray toPersonalizations(Mail email) {
        try {
            JSONArray jsonArray = new JSONArray();
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("to", toTos(email));
            jsonObject.put("subject", email.subject());
            jsonArray.put(jsonObject);
            return jsonArray;
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    private static JSONArray toTos(Mail email) {
        JSONArray tos = new JSONArray();
        email.recipients().map(SendGridMailService::toEmail).forEach(tos::put);
        return tos;
    }

    private static JSONObject toEmail(String address) {
        try {
            return new JSONObject().put("email", address);
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    interface SendGrid {

        @Headers("Content-Type: application/json")
        @RequestLine("POST /mail/send")
        void sendMail(String body);
    }
}

File: src/main/kotlin/io/github/binout/soccer/infrastructure/mail/SendGridMailService.kt
/*
 * Copyright 2016 Benoît Prioux
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.binout.soccer.infrastructure.mail

import feign.Feign
import feign.Headers
import feign.RequestLine
import io.github.binout.soccer.domain.MailService
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import java.util.*

@Component
class SendGridMailService : MailService {

    @Value("\${app.mail.sendgrid.url}")
    private val sendGridUrl: String? = null

    @Value("\${app.mail.sendgrid.api-key}")
    private val sendGridApiKey: String? = null

    override fun sendMail(mail: MailService.Mail) {
        if (!sendGridApiKey.isNullOrBlank()) {
            if (mail.hasRecipients()) {
                val jsonMail = toSendGridMail(mail)
                try {
                    Feign.builder()
                            .requestInterceptor { r -> r.header("Authorization", "Bearer $sendGridApiKey") }
                            .target(SendGrid::class.java, sendGridUrl)
                            .sendMail(jsonMail)
                } catch (e: Throwable) {
                    println(this.javaClass.name + " : " + "Cannot send mail to sendgrid api : $jsonMail")
                    println(this.javaClass.name + " : " + Arrays.toString(e.stackTrace))
                }

            }
        } else {
            println(this.javaClass.name + " : " + "No SendGrid Configuration")
        }
    }

    internal interface SendGrid {

        @Headers("Content-Type: application/json")
        @RequestLine("POST /mail/send")
        fun sendMail(body: String)
    }

    companion object {

        internal fun toSendGridMail(mail: MailService.Mail): String {
            try {
                val jsonObject = JSONObject()
                jsonObject.put("from", toEmail(mail.from()))
                jsonObject.put("personalizations", toPersonalizations(mail))
                jsonObject.put("content", toContent(mail))
                return jsonObject.toString()
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }

        }

        private fun toContent(mail: MailService.Mail): JSONArray {
            try {
                val jsonArray = JSONArray()
                val jsonObject = JSONObject()
                jsonObject.put("type", "text/html")
                jsonObject.put("value", mail.content())
                jsonArray.put(jsonObject)
                return jsonArray
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }

        }

        private fun toPersonalizations(mail: MailService.Mail): JSONArray {
            try {
                val jsonArray = JSONArray()
                val jsonObject = JSONObject()
                jsonObject.put("to", toTos(mail))
                jsonObject.put("subject", mail.subject())
                jsonArray.put(jsonObject)
                return jsonArray
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }

        }

        private fun toTos(mail: MailService.Mail): JSONArray {
            val tos = JSONArray()
            mail.recipients().map { toEmail(it) }.forEach { tos.put(it) }
            return tos
        }

        private fun toEmail(address: String): JSONObject {
            try {
                return JSONObject().put("email", address)
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }

        }
    }
}
